src/Libraries/DSCPython/DSCPythonCodeCompletionProviderCore.cs

```diff
@@ -0,0 +1,428 @@
+        }
+
+        private static PyScope CreateGlobalScope()
+        {
+            var scope = Py.CreateScope(globalScopeName);
+            // Allows discoverability of modules by inspecting their attributes
+            scope.Exec(@"
+import clr
+clr.setPreload(True)
+");
+
+            return scope;
+        }
+
+        private object ExecutePythonScriptCode(string code)
+        {
+            Python.Included.Installer.SetupPython().Wait();
+
+            if (!PythonEngine.IsInitialized)
+            {
+                PythonEngine.Initialize();
+                PythonEngine.BeginAllowThreads();
+            }
+
+            using (Py.GIL())
+            {
+                var result = Scope.Eval(code);
+                return CPythonEvaluator.OutputMarshaler.Marshal(result);
+            }
+        }
+
+        /// <summary>
+        /// List all of the members in a PythonObject
+        /// </summary>
+        /// <param name="pyObject">A reference to the module</param>
+        /// <param name="name">The name of the module</param>
+        /// <returns>A list of completion data for the module</returns>
+        private IEnumerable<Tuple<string, string, bool, ExternalCodeCompletionType>> EnumerateMembersForPyObj(object pyObject, string name)
+        {
+            var items = new List<Tuple<string, string, bool, ExternalCodeCompletionType>>();
+            var d = pyObject as PyObject;
+            if (d == null)
+            {
+                return items;
+            }
+
+            foreach (var member in d.Dir())
+            {
+                try
+                {
+                    var completionType = ExternalCodeCompletionType.Field;
+                    var attr = d.GetAttr(member.ToString()).ToString();
+                    if (attr.IndexOf(method, StringComparison.OrdinalIgnoreCase) >= 0 ||
+                        attr.IndexOf(inBuiltMethod, StringComparison.OrdinalIgnoreCase) >= 0)
+                    {
+                        completionType = ExternalCodeCompletionType.Method;
+                    }
+                    else if (attr.IndexOf("class", StringComparison.OrdinalIgnoreCase) >= 0)
+                    {
+                        completionType = ExternalCodeCompletionType.Class;
+                    }
+                    else if (attr.IndexOf("namespace", StringComparison.OrdinalIgnoreCase) >= 0 ||
+                             attr.IndexOf("module", StringComparison.OrdinalIgnoreCase) >= 0)
+                    {
+                        completionType = ExternalCodeCompletionType.Namespace;
+                    }
+
+                    items.Add(Tuple.Create((string)member.ToString(), name, false, completionType));
+                }
+                catch (Exception ex)
+                {
+                    Log(ex.ToString());
+                }
+            }
+
+            return items;
+        }
+
+        /// <summary>
+        /// Recursively lookup a variable in the _scope
```

Comment:
this method mentions recursion, but this does not look recursive, and all it seems to do is return false if the string contains a period?

Is it a base case of recursion maybe?

Reply:
renamed